### PR TITLE
drag&drop a VSIX file on the extensions viewlet

### LINKED ISSUE #12090: Implement VSIX drag and drop to extension viewlet
VSIX can now be installed via `code --install-extension extension.vsix` and via the action in the dropdown of the extension viewlet.

All that's left is the drag and drop onto the viewlet itself.

@bpasero Need your help here since the workbench seems to gobble the whole area for the drop feedback 👎

### PR DESCRIPTION
**Pull request by lszomoru**
This PR fixes #12090

Add the capability to drag&drop a VSIX file on the extensions viewlet and have it installed. While there is still room for improvement as well as some code duplication that I would like to explore removing, I want to get some early feedback on these changes. Thanks!

### CODE REVIEW

**Comment by lszomoru** [comment]
Tagging @joaomoreno and @sandy081 for review. Thanks!

**Comment by lszomoru** [comment]
I have updated the pull request to get rid of the code duplication (onDrop will invoke the existing install VSIX action), as well as I have fixed a bug due to which the `@installed` view was not refreshed after an extension was installed using the VSIX though the action or by dropping the VSIX on the extensions viewlet.

File: src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts [hunk 1]
```diff
@@ -617,6 +662,15 @@ export class ExtensionsViewPaneContainer extends ViewPaneContainer implements IE
 
 		this.notificationService.error(err);
 	}
+
+	private isSupportedDragElement(e: DragEvent): boolean {
+		if (e.dataTransfer) {
+			const typesLowerCase = e.dataTransfer.types.map(t => t.toLocaleLowerCase());
+			return typesLowerCase.indexOf('files') !== -1;
```
Comment (sandy081) [inline src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:669]: We shall we allow only `.vsix` files here.

**Comment by sandy081** [comment]
@lszomoru Thanks for the PR. I will take a look and get back to you.

BTW I updated the branch to make it easy for me to try out.

File: src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts [hunk 2]
```diff
@@ -431,6 +438,44 @@ export class ExtensionsViewPaneContainer extends ViewPaneContainer implements IE
 			}
 		}));
 
+		// Register DragAndDrop support
+		this._register(new DragAndDropObserver(this.root, {
```
Comment (sandy081) [inline src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:442]: image

Shall we have drop overlay only on extensions views excluding the search box ?

File: src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts [hunk 3]
```diff
@@ -396,8 +399,12 @@ export class ExtensionsViewPaneContainer extends ViewPaneContainer implements IE
 		addClass(parent, 'extensions-viewlet');
 		this.root = parent;
 
-		const header = append(this.root, $('.header'));
+		const overlay = append(this.root, $('.overlay'));
+		const overlayBackgroundColor = this.getColor(PANEL_DRAG_AND_DROP_BACKGROUND) ?? '';
```
Comment (sandy081) [inline src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:403]: Please use `SIDE_BAR_DRAG_AND_DROP_BACKGROUND` as this section is side bar

File: src/vs/workbench/contrib/extensions/browser/extensionsActions.ts [hunk 4]
```diff
@@ -2846,37 +2846,39 @@ export class InstallVSIXAction extends Action {
...
+		if (!vsixPaths) {
+			vsixPaths = await this.fileDialogService.showOpenDialog({
+				title: localize('installFromVSIX', "Install from VSIX"),
+				filters: [{ name: 'VSIX Extensions', extensions: ['vsix'] }],
+				canSelectFiles: true,
+				openLabel: mnemonicButtonLabel(localize({ key: 'installButton', comment: ['&& denotes a mnemonic'] }, "&&Install"))
+			});
+
+			if (!vsixPaths) {
 				return Promise.resolve();
```
Comment (sandy081) [inline src/vs/workbench/contrib/extensions/browser/extensionsActions.ts:2907]: Any empty return is enough as the method is `async`

**Review by sandy081 — COMMENTED** [review]
Changes look good to me except for some minor feedback (provided).

**Comment by sandy081** [comment]
>  as well as I have fixed a bug due to which the @installed view was not refreshed after an extension was installed using the VSIX though the action or by dropping the VSIX on the extensions viewlet.

Thanks for fixing this. Just to let you know there is a general bug that extension views do not get updated automatically when an extension is modified.

**Comment by lszomoru** [comment]
> > as well as I have fixed a bug due to which the @installed view was not refreshed after an extension was installed using the VSIX though the action or by dropping the VSIX on the extensions viewlet.
> 
> Thanks for fixing this. Just to let you know there is a general bug that extension views do not get updated automatically when an extension is modified.

@sandy081, do you have a link to the issue? Happy to take a look at it and submit a PR with the fix.

**Comment by sandy081** [comment]
> do you have a link to the issue? Happy to take a look at it and submit a PR with the fix.

Here is the issue - https://github.com/microsoft/vscode/issues/67603
I do not think the fix that is made here will help. We need a general fix - view (enabled/installed) should be updated if the view gets changed. It's not an issue that is suggested for contribution.

File: src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts (hunk 1 above)
Comment (lszomoru) [reply src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:669]: Yes, that is something that I can look at.

File: src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts (hunk 2 above)
Comment (lszomoru) [reply src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:442]: Based on my conversation with @joaomoreno, we wanted to keep things simple and that is why we are set up the drop target/drop overlay to be the shown on the whole extensions view. If you feel strongly about this, or there is a prior pattern not to show the drop overlay over the search box, I am happy to update it.

**Comment by joaomoreno** [comment]
@sandy081, I think you forgot this

**Comment by lszomoru** [comment]
Looks like `storage.ts` made it into this PR which is out of scope of this change. Working on removing it.

File: src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts (hunk 3 above)
Comment (lszomoru) [reply src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts]: Fixed.

File: src/vs/workbench/contrib/extensions/browser/extensionsActions.ts (hunk 4 above)
Comment (lszomoru) [reply src/vs/workbench/contrib/extensions/browser/extensionsActions.ts]: Fixed. I have also refactored the `run()` method to use `await`.

File: src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts (hunk 1 above)
Comment (lszomoru) [reply src/vs/workbench/contrib/extensions/browser/extensionsViewlet.ts:669]: I have researched this and based on my findings the file information (ex: path, extension) is not available in the `dragenter` and `dragover` event handlers. As that information is not available, we are not able to inspect the files that are being dragged over. This information is available in the drop event handler so, we are able to filter out files that are not `.vsix` files.

**Comment by lszomoru** [comment]
@sandy081, fixed up the pull request to revert the changes that were not in scope. Thanks for the feedback.

**Comment by lszomoru** [comment]
Abandoning the PR since my fork got into a funky state. Will re-create it from a clean fork.

**Comment by sandy081** [comment]
@joaomoreno nopes. Waiting for changes and discussing with @lszomoru over slack.

**Comment by lszomoru** [comment]
I have resubmitted these changes in a new pull request - #96907

**Pull request closed without merging**